### DISCUSSION TITLE
How to create a Directive for data fetching?

### DISCUSSION THREAD

**Discussion by rafaell-lycan** [MichalLytek/type-graphql#1434] [Q&A]
Hi y'all,

I want to ask for some guidance regarding adding new Directives to methods mainly to fetch data from other internal APIs.

The idea behind this is to use directives that match the HTTP verbs.

- httpGet
- httpPost
- etc.

**Note:** It could be just `@rest` with the `method` as an optional parameter.

Ideally it would use a pattern like so:
```gql
# a simple example
@httpGet("api.example.com/v2/resource/$id")

# or even something more advanced
@httpGet(url: "api.example.com/v2/resource/$id")
```

Any ideas on similar solutions or how to develop it?

**Comment by carlocorradini** [comment]
You can start here: https://the-guild.dev/graphql/tools/docs/schema-directives

And integrate with TypeGraphQL following the example: https://github.com/carlocorradini/graphql-auth-directive/blob/main/examples/typegraphql/index.ts

**Reply by carlocorradini** [reply]
@rafaell-lycan Solved it?

**Reply by rafaell-lycan** [reply]
@carlocorradini not entirely.

I guess it would also be possible to add different properties such as `map` or `resolve` to specify which part of the JSON response you'll get to match the Schema and its Types right?

**Reply by carlocorradini** [reply]
@rafaell-lycan Example?

**Reply by rafaell-lycan** [reply]
For example getting a specific field from a JSON file using JSON Path/XPath, changing the HTTP method, etc.

```gql
type Query {
  # ...
  form(id: ID!): Form! @httpGet(api: "api.example.com/v2/resource/$id", resolve: "$.data")
  timeToCompleteByFormId(id: ID!): Int @httpGet(api: "api.example.com/v2/resource/$id", method: "POST")
  formFeatures(id: ID!): [String] @httpGet(api: "api.example.com/v2/resource/$id", resolve: "$.features[*]")
  # ...
}

**Reply by carlocorradini** [reply]
@rafaell-lycan 
I think `@http` should be a better name for the directive :)

So you want to perform an HTTP request and pass the value in the resolver? Or as a result?

**Reply by rafaell-lycan** [reply]
The general idea would be not writing a resolver, meaning this would play more as a DSL for developers without needing to touch JS/TS code.

I was thinking about just `@http` or even `@request` :-)

**Reply by carlocorradini** [reply]
And then store the value of the request to a variable?
